File: scala-design-patterns/src/main/scala/StructuralPatterns/Composite/Composite.scala
import scala.collection.mutable.ListBuffer

trait Component[T] {
  def operation(): T
}

class Leaf (value: Int) extends Component[Int] {
  override def operation(): Int = this.value
}

class Composite extends Component[Int] {
  private val children = new ListBuffer[Component[Int]]()
  def +=(child: Component[Int]): Unit = children += child
  def -=(child: Component[Int]): Unit = children -= child
  override def operation(): Int = children.map(_.operation()).sum
}

object Composite {
  def main(args: Array[String]): Unit = {
    var root = new Composite()
    root += new Leaf(10)
    root += new Leaf(20)
    
    var child = new Composite()
    child += new Leaf(30)
    child += new Leaf(40)

    root += child
    
    println(root.operation())
  }
}

File: python-design-patterns/structural_patterns/composite/composite.py
class Component:
    def __init__(self):
        self.children = []

    def add(self, child):
        self.children.append(child)

    def remove(self, child):
        self.children.remove(child)

    def operation(self):
        pass


class Leaf(Component):
    def __init__(self, value):
        super().__init__()
        self.value = value

    def operation(self):
        return self.value


class Composite(Component):
    def __init__(self):
        super().__init__()

    def operation(self):
        return sum([child.operation() for child in self.children])


def main():
    root = Composite()
    root.add(Leaf(10))
    root.add(Leaf(20))

    child = Composite()
    child.add(Leaf(30))
    child.add(Leaf(40))

    root.add(child)

    print(root.operation())


if __name__ == "__main__":
    main()
